kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java

```diff
@@ -387,18 +392,30 @@ engine, getLogReplayFiles(logSegment), SET_TRANSACTION_READ_SCHEMA, Optional.emp
    */
   private Map<String, DomainMetadata> loadDomainMetadataMap(Engine engine) {
     // First try to load from CRC info if available
-    Optional<CRCInfo> currentCrcInfo = getCurrentCrcInfo();
-    if (currentCrcInfo.isPresent() && currentCrcInfo.get().getDomainMetadata().isPresent()) {
-      return currentCrcInfo.get().getDomainMetadata().get().stream()
+    Optional<CRCInfo> lastSeenCrcInfoOpt = crcInfoContext.getLastSeenCrcInfo();
+    if (!lastSeenCrcInfoOpt.isPresent()
+        || !lastSeenCrcInfoOpt.get().getDomainMetadata().isPresent()) {
+      logger.info("No domain metadata available in CRC info, loading from log");
+      return loadDomainMetadataMapFromLog(engine, Optional.empty());
+    }
+    CRCInfo lastSeenCrcInfo = lastSeenCrcInfoOpt.get();
+    if (lastSeenCrcInfo.getVersion() == logSegment.getVersion()) {
+      return lastSeenCrcInfo.getDomainMetadata().get().stream()
           .collect(Collectors.toMap(DomainMetadata::getDomain, Function.identity()));
     }
-    // TODO https://github.com/delta-io/delta/issues/4454: Incrementally load domain metadata from
-    // CRC when current CRC is not available.
-    // Fall back to loading from the log
-    logger.info("No domain metadata available in CRC info, loading from log");
-    return loadDomainMetadataMapFromLog(engine).entrySet().stream()
-        .filter(entry -> !entry.getValue().isRemoved())
-        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    Map<String, DomainMetadata> finalDomainMetadataMap =
+        loadDomainMetadataMapFromLog(engine, Optional.of(lastSeenCrcInfo.getVersion() + 1));
+    lastSeenCrcInfo
+        .getDomainMetadata()
+        .get()
+        .forEach(
+            domainMetadataInCrc -> {
+              if (!finalDomainMetadataMap.containsKey(domainMetadataInCrc.getDomain())) {
+                finalDomainMetadataMap.put(domainMetadataInCrc.getDomain(), domainMetadataInCrc);
+              }
```

Comment:
nit - can you maybe add a clarifying comment why this works? it was non-trivial for me to understand that if a domain has been removed, it would also be present in the map (with it removed) & thus won't be added. maybe just a TLDR how to combine the two maps would be helpful